### PR TITLE
[dependentAPI operator] Add logic to dynamically retrieve the API endpoint from the custom resource & Update status as True

### PR DESCRIPTION
**Pull request by aankursarkar**
Enhancement: Dynamically Retrieve API Endpoint from Custom Resource

This pull request introduces the following changes:

    Dynamic API Endpoint Retrieval:
        Add logic to dynamically retrieve the API endpoint from the custom resource.
        Update implementation status as True.

    Chart Versions Update:
        Update the Canvas and dependentAPI Charts versions to reflect the new changes.

    RBAC Helm Template Update:
        Modify the DependentAPI RBAC Helm template to allow access to API custom resources.

    Docker Image Registry:
        Use a personal DockerHub registry in the charts to store the updated dependent Docker image due to restricted access over the tmforumodacanvas repository.

### CODE REVIEW

File: charts/canvas-oda/values.yaml [hunk 1]
```diff
@@ -131,7 +131,7 @@ controller:
     
 dependentapi-simple-operator:
   enabled: true
-  image: mtr.devops.telekom.de/magenta_canvas/public:dependentapi-simple-operator-0.1.0
+  image: ankursarkar/dependentapi-simple-operator:0.2.0-rc
```
Comment (ferenc-hechler) [inline charts/canvas-oda/values.yaml:134]: For the pull request to master branch, we should use a fixed version, not a release-candidate (*-rc), which might be overwritten by some pipeline.

File: charts/dependentapi-simple-operator/templates/rbac.yaml [hunk 2]
```diff
@@ -39,6 +39,10 @@ rules:
     resources: [components]
     verbs: [list, watch, get, patch]
 
+  - apiGroups: [oda.tmforum.org]
+    resources: [apis]
+    verbs: [list, watch, get, patch]
```
Comment (ferenc-hechler) [inline charts/dependentapi-simple-operator/templates/rbac.yaml:44]: Is "patch" really needed? I think APIs should be accessed read-only.

File: charts/dependentapi-simple-operator/templates/rbac.yaml [hunk 3]
```diff
@@ -92,6 +96,11 @@ rules:
     resources: [components]
     verbs: [list, watch, patch, get]
 
+  # Application: watching & handling for the custom resource we declare.
+  - apiGroups: [oda.tmforum.org]
+    resources: [apis]
+    verbs: [list, watch, get, patch]
```
Comment (ferenc-hechler) [inline charts/dependentapi-simple-operator/templates/rbac.yaml:102]: "patch" needed?

File: charts/dependentapi-simple-operator/values.yaml [hunk 4]
```diff
@@ -1,3 +1,3 @@
-image: mtr.devops.telekom.de/magenta_canvas/public:dependentapi-simple-operator-0.1.0
+image: ankursarkar/dependentapi-simple-operator:0.2.0-rc
```
Comment (ferenc-hechler) [inline charts/dependentapi-simple-operator/values.yaml:1]: avoid using release-candidate versions "-rc"

Comment (aankursarkar) [reply charts/dependentapi-simple-operator/values.yaml:1]: Will update the image without "-rc"

File: charts/dependentapi-simple-operator/templates/rbac.yaml (hunk 3 above)
Comment (aankursarkar) [reply charts/dependentapi-simple-operator/templates/rbac.yaml:102]: API PATCH is not needed for dependent API.
I will remove the PATCH from the list and raise a new PR.

File: source/operators/dependentApiSimpleOperator/docker/dependentApiSimpleOperator.py [hunk 5]
```diff
@@ -63,6 +65,50 @@ def safe_get(default_value, dictionary, *paths):
...
+    except ApiException as e:
+        if e.status == HTTP_NOT_FOUND:
+            logger.error(
+                f"setDependentAPIStatus: dependentapi {depapi_namespace}:{depapi_name} not found"
+            )
+        else:
+            raise kopf.TemporaryError(
+                f"setDependentAPIStatus: Exception in get_namespaced_custom_object: {e.body}"
+            )           
+def get_expapi():
```
Comment (ferenc-hechler) [inline source/operators/dependentApiSimpleOperator/docker/dependentApiSimpleOperator.py:84]: This file was formatted using `black` with the default settings. 
I would suggest, to do this again.

Black can be installed using pip.
To format code, just change into the folder and execute

```
$ black .
reformatted ~/git/oda-canvas/source/operators/dependentApiSimpleOperator/docker/dependentApiSimpleOperator.py

All done! ✨ 🍰 ✨
1 file reformatted, 1 file left unchanged.
````

Comment (aankursarkar) [reply source/operators/dependentApiSimpleOperator/docker/dependentApiSimpleOperator.py:87]: As suggested, run the black on the dependent API operator
image

image

File: charts/canvas-oda/values.yaml (hunk 1 above)
Comment (aankursarkar) [reply charts/canvas-oda/values.yaml]: -rc has been remove from the build.sh & build.cmd

File: charts/dependentapi-simple-operator/templates/rbac.yaml (hunk 2 above)
Comment (aankursarkar) [reply charts/dependentapi-simple-operator/templates/rbac.yaml]: revoke Patch access for API CR

**Review by ferenc-hechler — APPROVED** [review]

**Comment by brian-burton** [comment]
@aankursarkar, this PR has conflicts, please can you check?

**Comment by aankursarkar** [comment]
> @aankursarkar, this PR has conflicts, please can you check?

@brian-burton,
Conflicts has been resolve, Could you please have a look on this PR.
Thank you.!!

**Comment by aankursarkar** [comment]
Due to rights issue, I have to use personal docker registry to keep the dependentAPI docker image. But this image need to be push to the "tmforumorg" registry.

**Comment by brian-burton** [comment]
> Due to rights issue, I have to use personal docker registry to keep the dependentAPI docker image. But this image need to be push to the "tmforumorg" registry.

OK, @aankursarkar, understood. We're trying to sort that out at the moment so that we can build images more effectively.

**Review by ferenc-hechler — APPROVED** [review]
Rereviewed the PR. Looks all good!  :-)

**Pull request merged**